### ISSUE TITLE
cyrpto/x509: ability to add custom curve when parsing X509 certificate

### ISSUE THREAD

**Issue by trung** [golang/go#32874]
Per https://github.com/golang/go/issues/26776, using a third party library for custom curve is advised. 

However, when parsing x509 certificate (`x509.ParseCetificate()`), it is not possible to supply custom curve.

My proposal is to offer a configuration that can be used to supply a function to return `elliptic.Curve` from `asn1.ObjectIdentifier` to complement the default `namedCurveFromIOD()`

**Comment by trung** [comment]
From the current flow:
https://github.com/golang/go/blob/fbde753a58e286c405a04388816ed044486151bb/src/crypto/x509/x509.go#L519-L553

I think of injecting custom functions something like:

```go
type OIDToCurveFunc func(oid asn1.ObjectIdentifier) elliptic.Curve
type CurveToOIDFunc func(curve elliptic.Curve) (asn1.ObjectIdentifier, bool)

var (
	CustomIOIDToCurveFunc OIDToCurveFunc
	CustomCurveToOIDFunc CurveToOIDFunc
)

func namedCurveFromOID(oid asn1.ObjectIdentifier) elliptic.Curve {
	switch {
	case oid.Equal(oidNamedCurveP224):
		return elliptic.P224()
	case oid.Equal(oidNamedCurveP256):
		return elliptic.P256()
	case oid.Equal(oidNamedCurveP384):
		return elliptic.P384()
	case oid.Equal(oidNamedCurveP521):
		return elliptic.P521()
	}
	if CustomIOIDToCurveFunc != nil {
		return CustomIOIDToCurveFunc(oid)
	}
	return nil
}

func oidFromNamedCurve(curve elliptic.Curve) (asn1.ObjectIdentifier, bool) {
	switch curve {
	case elliptic.P224():
		return oidNamedCurveP224, true
	case elliptic.P256():
		return oidNamedCurveP256, true
	case elliptic.P384():
		return oidNamedCurveP384, true
	case elliptic.P521():
		return oidNamedCurveP521, true
	}
	if CustomCurveToOIDFunc != nil {
		return CustomCurveToOIDFunc(curve)
	}
	return nil, false
}

```

**Comment by rsc** [comment]
Globals are probably not the right answer here. Perhaps there should be an x509.Parser that can be configured with extra hooks, a bit like there is an xml.Decoder to configure the XML parser.

Or perhaps unrecognized curves should be exposed in some general way and client code can just post-process.

/cc @FiloSottile

**Comment by trung** [comment]
Agreed with globals. There might be breaking changes once introducing `x509.Parser` as `x509.ParseCertifcate()` is used in number of places.

**Comment by FiloSottile** [comment]
crypto/x509 primarily serves the WebPKI, and there is no use of custom curves there. X.509 is a sprawling standard, so it's critical for the crypto/x509 package to stay focused on its use case.

I will keep this open to collect feedback about use cases, as things of course change over time, but we are not going to support this for now.

**Comment by Joh4nnesHartl** [comment]
Is there any update on this, especially when 1.20 will deprecate custom curve implementations? Still no way to provide a workaround? 

cc  @FiloSottile

**Comment by ZackaryWelch** [comment]
Strange to exclude over half of support openssl curves. Running into this myself and I am having to inject the four Brainpool OIDs and their curves from another library. It'd be fairly trivial to support these, either the regular or both regular and twisted curves.

**Comment by eriklupander** [comment]
Any news on this issue? 

I would like to support certificates signed with ECDSA keys using brainpool curves in the application I'm working on. I'd prefer to do that without relying on OpenSSL, or by parsing the cert bytes myself in order to get hold of the artifacts needed to use a 3rd party module with brainpool support.

Would really appreciate support for custom curves.

**Comment by jhollmannk** [comment]
We have the same problem here with the brainpool curves. Yes, it's a german thing but unfortunately we have no choice here as we just have to support this.

Does anyone have an example how to even do this? I'm at a loss. Do I have to replace the x509 package (or the methods I need to read a certificate) and do the same with the tls package when it comes to a TLS handshake?

@ZackaryWelch mentioned that he has to inject the brainpoool curves. I would like to know how and where to inject them.

**Comment by kostas-zealid** [comment]
We are interested in this as well. ICAO 9303 (which defines eMRTDs including NFC data structure and signatures inside all worldwide travel documents incl. passwords and ID cards) mandates custom curves with explicit parameters as part of Document Signing Certificate inside eMRTD NFC chips.

I wonder what is the best way forward for us given that we'd like to consider the option of validating these crypto signatures using native Go crypto.

**Comment by kwe** [comment]
Just wondering if anyone has found a solution to the issue with eMRTDs and using the native Go crypto?

**Comment by Joh4nnesHartl** [comment]
> Just wondering if anyone has found a solution to the issue with eMRTDs and using the native Go crypto?

Fork the existing go-crypto and add your OIDs, the curves used also need to implement the `elliptic.Curve` interface.

The only functions I had to adjust were `namedCurveFromOID` and `oidFromNamedCurve` inside `x509/x509.go`

for future versions you could also prepare a patch file for these changes.

**Comment by Lazybin** [comment]
Other languages can parse these curves, but Go can't. It's really uncomfortable to use it now.

**Comment by septs** [comment]
I really need brainpool elliptic curves support now.

**Comment by ZackaryWelch** [comment]
This is how I had done it in the past, editing `x509.go` to include the curves

```
  # Inject brainpool OIDs to x509.go
WORKDIR /usr/local/go/src/crypto/x509
RUN sed -i -e 's/{1, 3, 132, 0, 35}/{1, 3, 132, 0, 35}\noidNamedCurveP224Brainpool = asn1.ObjectIdentifier{1,3,36,3,3,2,8,1,1,5}\n\toidNamedCurveP256Brainpool = asn1.ObjectIdentifier{1,3,36,3,3,2,8,1,1,7}\n\toidNamedCurveP384Brainpool = asn1.ObjectIdentifier{1,3,36,3,3,2,8,1,1,11}\n\toidNamedCurveP512Brainpool = asn1.ObjectIdentifier{1,3,36,3,3,2,8,1,1,13}\n/g' x509.go && \

  sed -i -e 's/return elliptic.P521()/return elliptic.P521()\ncase oid.Equal(oidNamedCurveP224Brainpool):\n\treturn brainpool.P224r1()\n\tcase oid.Equal(oidNamedCurveP256Brainpool):\n\treturn brainpool.P256r1()\n\tcase oid.Equal(oidNamedCurveP384Brainpool):\n\treturn brainpool.P384r1()\n\tcase oid.Equal(oidNamedCurveP512Brainpool):\n\treturn brainpool.P512r1()\n/g' x509.go && \

  sed -i -e 's|\"crypto/sha512\"|"crypto/sha512"\n  brainpool '\"github.com/go-compile/rome/brainpool/bcurves\"' \n|g' x509.go && \
  go get github.com/go-compile/rome && go mod tidy
```

**Comment by Joh4nnesHartl** [comment]
with the new rewrite of the x509 package in go1.23 I think this opens more possibilities.

**Comment by t4w4n3** [comment]
I need to parse x509 der certificates with a brainpool curve in Golang.

Why ? Because the product I'm contributing to follow the [ANSSI](https://en.wikipedia.org/wiki/Agence_nationale_de_la_s%C3%A9curit%C3%A9_des_syst%C3%A8mes_d%27information) recomandations.

On this page : [https://cyber.gouv.fr/publications/corpus-documentaire-ipsec-dr-destination-des-industriels-version-10](https://cyber.gouv.fr/publications/corpus-documentaire-ipsec-dr-destination-des-industriels-version-10)

The first document gives recomandations on IPSEC DR : [https://cyber.gouv.fr/sites/default/files/document/note_cryptographique_ipsec_dr_v1.0.pdf](https://cyber.gouv.fr/sites/default/files/document/note_cryptographique_ipsec_dr_v1.0.pdf)

Page 6, chapter 2 the two curves are :
* BrainpoolP256r1
* secp256r1

The SecNumCloud qualification requires following the ANSSI recomandations.

For Golang to continue to be chosen as a SecNumCloud compliant language in Europe, then it needs to include more curves, **specifically brainpool**, or provide a convenient way to use custom curves.
